app/routes.py

```diff
@@ -1 +1,206 @@
+
+from flask import Blueprint, jsonify, abort, make_response, request
+import datetime
+import os
+import requests
+
+task_bp = Blueprint("tasks", __name__, url_prefix="/tasks")
+goal_bp = Blueprint("goals", __name__, url_prefix="/goals")
+
+
+@task_bp.route("", methods=["POST"])
+def make_a_task():
+    try:
+        request_body = request.get_json()
+
+        new_task = Task.from_dict(request_body)
+
+        db.session.add(new_task)
+        db.session.commit()
+
+        return {
+        "task": new_task.to_dict()}, 201
+    except KeyError as error:
+        return {"details": "Invalid data"}, 400
+
+@task_bp.route("", methods=["GET"])
+def get_all_tasks():
+    title_query = request.args.get("title")
+    desc_query = request.args.get("description")
+    sort_query = request.args.get("sort")
+
+    query_map = {
+        Task.query.filter_by(title = title_query): title_query,
+        Task.query.filter_by(description = desc_query):desc_query, 
+        Task.query.order_by(Task.title): sort_query == "asc",
+        Task.query.order_by(Task.title.desc()): sort_query == "desc"
+    }
+    tasks = []
+    task_response = []
+
+    for key, value in query_map.items():
+        if value:
+            tasks = key
+    if not tasks:
+        tasks = Task.query.all()
+    task_response = [task.to_dict() for task in tasks]
+    return jsonify(task_response)
+
+def validate_model(model_class, model_id):
+    try:
+        model_id = int(model_id)
+    except:
+        abort(make_response({"message": f"{model_id} is not a valid type ({type(model_id)}. Must be an integer)"}, 400))
+
+    model = model_class.query.get(model_id)
+
+    if not model:
+        abort(make_response({"message": f"{model_class.__name__} {model_id} does not exist."}, 404))
+    
+    return model
+
+@task_bp.route("/<task_id>", methods=["GET"])
+def get_one_task(task_id):
+    task = validate_model(Task,task_id)
+    if not task.goal:
+        return {"task" :task.to_dict()}, 200
+    else:
+        return {"task" :task.goal_to_dict()}, 200
+
+@task_bp.route("/<task_id>", methods=["PUT"])
+def update_task(task_id):
+    task = validate_model(Task,task_id)
+
+    request_body = request.get_json()
+    
+    try:
+        task.title = request_body["title"]
+        task.description = request_body["description"]
+    except:
+        return {"details": "Invalid data"}, 404
+
+    db.session.commit()
+    return  {"task": task.to_dict()}, 200
+
+@task_bp.route("/<task_id>", methods=["DELETE"])
+def delete_task(task_id):
+    task = validate_model(Task,task_id)
+    db.session.delete(task)
+    db.session.commit()
+    return  {
+        "details": f"Task {task_id} \"{task.title}\" successfully deleted"
+    }, 200
+
+@task_bp.route("/<task_id>/mark_complete", methods=["PATCH"])
+def mark_title_as_complete(task_id):
+    task = validate_model(Task,task_id)
+    task.completed_at = datetime.datetime.now()
+    db.session.commit()
+
+    post_to_slack(task)
```

Comment:
Nice work on having your Slack post sent after the logic of marking a task complete! We don't want to send out any false positive alerts just in case our logic fails during the update!